packages/nextra-theme-docs/src/misc/theme.tsx

```diff
@@ -89,33 +85,23 @@ if (typeof window !== 'undefined') {
 }
 
 // Anchor links
-const HeaderLink = ({
-  tag: Tag,
+const createHeaderLink = (Tag: `h${2 | 3 | 4 | 5 | 6}`) => function HeaderLink({
   children,
-  id,
-  context,
-  withObserver = true,
+  id: slug,
   ...props
 }: {
   tag: any
-  children: any
+  children: ReactNode
   id: string
-  context: { index: number }
-  withObserver?: boolean
-}) => {
+}): ReactElement {
   setActiveAnchor = useActiveAnchorSet()
   const obRef = useRef<HTMLSpanElement>(null)
 
-  const slug = id
-  const anchor = <span className="subheading-anchor" id={slug} ref={obRef} />
-
-  const index = context.index++
-
   useEffect(() => {
     const ref = obRef
     if (!ref.current) return
 
-    slugs.set(ref.current, [slug, index])
+    slugs.set(ref.current, [slug, Number(Tag[1]) - 1])
```

Comment:
I don't think this will work. The current `index` represents the index of the heading over all headings in the document, for example:

```
h1 → index = 0
h2 → index = 1
h2 → index = 2
h3 → index = 3
h2 → index = 4
```

Reply:
This refactoring matches current behavior, but now it’s not mutate `context` object
https://github.com/shuding/nextra/blob/07e4732f447e392319150f03354d6a1e365a81c1/packages/nextra-theme-docs/src/misc/theme.tsx#L112
h1 doesn’t exist
h2 context.index=0, but index is 1 (context.index++)

Reply:
@shuding thank you! you were totally right!
So let's export `getComponents` function instead of `components` object

Reply:
Sounds good! I have to say that the `context.index++` thing wasn't an elegant solution, but it's very tricky in React. If possible we can brainstorm to find another work around.

Reply:
Sure, for now it’s ok, as it’s works, but for the future we can rethink 👍